### PR TITLE
Handle touchcancel events

### LINKED ISSUE #1286: touchcancel
From here: https://plus.google.com/u/0/+RickByers/posts/Ny6ZXuzWdN5
The link deals mostly with scrolling, but it applies to anyone listening to `touchend`.  

There are multiple areas in the code where we listen to `touchend` events, but we never subscribe to `touchcancel`.  This creates some nasty bugs where our `touchend` code never gets called (which I've seen first hand on Android but didn't know what the problem was until I stumbled across this link).  We need to subscribe to both and terminate our touch event accordingly.  

CC @emackey @bagnell

### PR DESCRIPTION
**Pull request by twpayne**
This is an attempt to fix #1286.

The code isn't ready to be merged yet, but this PR is opened for discussion on whether this is the correct approach (or not).

cc @emackey

### CODE REVIEW

File: Source/Core/ScreenSpaceEventType.js [hunk 1]
```diff
@@ -145,7 +145,15 @@ define([
...
-        PINCH_MOVE : 19
+        PINCH_MOVE : 19,
+
+        /**
+         * Represents a cancellation of a two-finger event on a touch surface.
+         *
+         * @type {Number}
+         * @constant
+         */
+        PINCH_CANCEL : 20
```
Comment (emackey) [inline Source/Core/ScreenSpaceEventType.js]: We don't need any user-defined action for a cancel event.  The actual cancel event is extremely rare, for example when putting an Android device to sleep in the middle of a touch action.  I would just fire `PINCH_END` and not define a new separate `PINCH_CANCEL`.  It's not like the camera is going to undo the pinch-zooming that has happened just prior to the cancel.

File: Source/Core/ScreenSpaceEventHandler.js [hunk 2]
```diff
@@ -91,7 +91,8 @@ define([
             registerListener(screenSpaceEventHandler, 'mouseup', alternateElement, handleMouseUp);
             registerListener(screenSpaceEventHandler, 'mousemove', alternateElement, handleMouseMove);
             registerListener(screenSpaceEventHandler, 'touchstart', element, handleTouchStart);
-            registerListener(screenSpaceEventHandler, 'touchend', alternateElement, handleTouchEnd);
+            registerListener(screenSpaceEventHandler, 'touchend', alternateElement, handleTouchEndOrCancel);
+            registerListener(screenSpaceEventHandler, 'touchcancel', alternateElement, handleTouchEndOrCancel);
```
Comment (emackey) [inline Source/Core/ScreenSpaceEventHandler.js]: We listen for `touchcancel` but not `pointercancel` here.  We should do both.

File: Source/Scene/CameraEventAggregator.js [hunk 3]
```diff
@@ -104,6 +104,12 @@ define([
                 movement.angleAndHeight.startPosition.x = -prevAngle * canvas.clientWidth / 12;
             }
         }, ScreenSpaceEventType.PINCH_MOVE, modifier);
+
+        aggregator._eventHandler.setInputAction(function() {
+            aggregator._buttonsDown = Math.max(aggregator._buttonsDown - 1, 0);
+            isDown[key] = false;
+            releaseTime[key] = new Date();
+        }, ScreenSpaceEventType.PINCH_CANCEL, modifier);
```
Comment (emackey) [inline Source/Scene/CameraEventAggregator.js]: This code is identical to the code for `PINCH_END`.  See my other comment.

**Comment by twpayne** [comment]
Thanks for the review/feedback. All comments addressed.

The tl;dr is that we now treat `pointercancel` as `pointerup` everywhere, and `touchcancel` as `touchend` everywhere.

Note that `Animation` and `Timeline` do not currently listen for *any* `pointer` events. I'd suggest adding these in a separate PR.

**Comment by hpinkos** [comment]
> Note that Animation and Timeline do not currently listen for any pointer events. I'd suggest adding these in a separate PR.

@twpayne can you write up an issue for this please?

File: Source/Core/ScreenSpaceEventType.js (hunk 1 above)
Comment (twpayne) [inline Source/Core/ScreenSpaceEventType.js]: Done.

File: Source/Core/ScreenSpaceEventHandler.js (hunk 2 above)
Comment (twpayne) [inline Source/Core/ScreenSpaceEventHandler.js]: Done.

File: Source/Scene/CameraEventAggregator.js (hunk 3 above)
Comment (twpayne) [inline Source/Scene/CameraEventAggregator.js]: Done.

**Pull request merged**